File: domain.guice/src/test/java/org/openmedstack/domain/guice/OptimisticEventStreamTest.java
package org.openmedstack.domain.guice;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.openmedstack.eventstore.CommitAttempt;
import org.openmedstack.eventstore.Commit;
import org.openmedstack.eventstore.ICommitEvents;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class OptimisticEventStreamTest {

    @Before
    public void setUp() throws Exception {
    }

    @After
    public void tearDown() throws Exception {
    }

    @Test
    public void canCreateInstance() {
        var instance = OptimisticEventStream.Create("bucket", "stream", new ICommitEvents() {
            @Override
            public CompletableFuture<Iterable<Commit>> getFrom(String bucketId, String streamId, int minRevision, int maxRevision) {
                return null;
            }

            @Override
            public CompletableFuture<Commit> commit(CommitAttempt attempt) {
                return null;
            }
        });
        Assert.assertNotNull(instance);
    }

    @Test
    public void canCreateFutureInstance() {
        var future = OptimisticEventStream.Create("bucket", "stream", new ICommitEvents() {
            @Override
            public CompletableFuture<Iterable<Commit>> getFrom(String bucketId, String streamId, int minRevision, int maxRevision) {
                return CompletableFuture.completedFuture(List.of(new Commit[0]));
            }

            @Override
            public CompletableFuture<Commit> commit(CommitAttempt attempt) {
                return null;
            }
        }, 0, Integer.MAX_VALUE);
        try {
            var instance = future.get();

            Assert.assertNotNull(instance);
        } catch (Exception e) {
            Assert.fail();
        }
    }

    public void testTestCreate1() {
    }

    public void testBucketId() {
    }

    public void testGetStreamId() {
    }

    public void testGetStreamRevision() {
    }

    public void testGetCommitSequence() {
    }

    public void testGetCommittedEvents() {
    }

    public void testGetCommittedHeaders() {
    }

    public void testGetUncommittedEvents() {
    }

    public void testGetUncommittedHeaders() {
    }

    public void testAdd() {
    }

    public void testCommitChanges() {
    }

    public void testClearChanges() {
    }

    public void testClose() {
    }
}

File: domain.guice/src/test/kotlin/org/openmedstack/domain/guice/OptimisticEventStreamTest.kt
package org.openmedstack.domain.guice

import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.openmedstack.domain.guice.OptimisticEventStream.Companion.create
import org.openmedstack.eventstore.Commit
import org.openmedstack.eventstore.CommitAttempt
import org.openmedstack.eventstore.CommitImpl
import org.openmedstack.eventstore.ICommitEvents
import java.util.concurrent.CompletableFuture

class OptimisticEventStreamTest {
    @Before
    @Throws(Exception::class)
    fun setUp() {
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
    }

    @Test
    fun canCreateInstance() {
        val instance = create("bucket", "stream", object : ICommitEvents {
            override fun getFrom(
                bucketId: String?,
                streamId: String,
                minRevision: Int,
                maxRevision: Int
            ): CompletableFuture<Iterable<Commit>> {
                return CompletableFuture.completedFuture(listOf<Commit>())
            }

            override fun commit(attempt: CommitAttempt): CompletableFuture<Commit> {
                return CompletableFuture.completedFuture(
                    CommitImpl.create(attempt)
                )
            }
        })
        Assert.assertNotNull(instance)
    }

    @Test
    fun canCreateFutureInstance() {
        val future = create("bucket", "stream", object : ICommitEvents {
            override fun getFrom(
                bucketId: String?,
                streamId: String,
                minRevision: Int,
                maxRevision: Int
            ): CompletableFuture<Iterable<Commit>> {
                return CompletableFuture.completedFuture(listOf())
            }

            override fun commit(attempt: CommitAttempt): CompletableFuture<Commit> {
                return CompletableFuture.completedFuture(CommitImpl.create(attempt))
            }
        }, 0, Int.MAX_VALUE)
        try {
            val instance = future.get()
            Assert.assertNotNull(instance)
        } catch (e: Exception) {
            Assert.fail()
        }
    }
}
